// lotus-typescript/dist/data-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateEventType = exports.REQUEST_URLS = exports.REQUEST_TYPES = void 0;
var REQUEST_TYPES;
(function (REQUEST_TYPES) {
    REQUEST_TYPES["GET"] = "GET";
    REQUEST_TYPES["POST"] = "POST";
    REQUEST_TYPES["PATCH"] = "PATCH";
})(REQUEST_TYPES = exports.REQUEST_TYPES || (exports.REQUEST_TYPES = {}));
exports.REQUEST_URLS = {
    GET_CUSTOMERS: "/api/customers/",
    CREATE_CUSTOMERS: "/api/customers/",
    CREATE_BATCH_CUSTOMERS: "/api/batch_create_customers/",
    GET_CUSTOMER_DETAIL: (customerId) => `/api/customers/${customerId}/`,
    CREATE_SUBSCRIPTION: "/api/subscriptions/",
    CANCEL_SUBSCRIPTION: (subscriptionId) => `/api/subscriptions/${subscriptionId}/`,
    CHANGE_SUBSCRIPTION: (subscriptionId) => `/api/subscriptions/${subscriptionId}/`,
    GET_ALL_SUBSCRIPTIONS: "/api/subscriptions/",
    GET_SUBSCRIPTION_DETAILS: (subscriptionId) => `/api/subscriptions/${subscriptionId}/`,
    GET_ALL_PLANS: "/api/plans/",
    GET_CUSTOMER_ACCESS: "/api/customer_access/",
    TRACK_EVENT: "/api/track/",
};
var ValidateEventType;
(function (ValidateEventType) {
    ValidateEventType["createCustomer"] = "createCustomer";
    ValidateEventType["trackEvent"] = "trackEvent";
    ValidateEventType["customerDetails"] = "customerDetails";
    ValidateEventType["createSubscription"] = "createSubscription";
    ValidateEventType["cancelSubscription"] = "cancelSubscription";
    ValidateEventType["changeSubscription"] = "changeSubscription";
    ValidateEventType["subscriptionDetails"] = "subscriptionDetails";
    ValidateEventType["customerAccess"] = "customerAccess";
    ValidateEventType["createCustomersBatch"] = "createCustomersBatch";
})(ValidateEventType = exports.ValidateEventType || (exports.ValidateEventType = {}));
